### PR TITLE
Channel task cancelation

### PR DESCRIPTION
**Pull request by JonathanHenson**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/event_loop.c [hunk 1]
```diff
@@ -244,6 +244,12 @@ void aws_event_loop_schedule_task_future(
     event_loop->vtable->schedule_task_future(event_loop, task, run_at_nanos);
 }
 
+void aws_event_loop_cancel_task(struct aws_event_loop *event_loop, struct aws_task *task) {
```
Comment (justinboswell) [inline source/event_loop.c]: Documentation says this must be called from event_loop's thread, should we assert on that here?

File: include/aws/io/channel.h [hunk 2]
```diff
@@ -59,6 +60,9 @@ struct aws_channel {
     struct aws_shutdown_notification_task shutdown_notify_task;
     aws_channel_on_shutdown_completed_fn *on_shutdown_completed;
     void *shutdown_user_data;
+    struct aws_linked_list pending_tasks;
+    struct aws_mutex x_thread_task_lock;
```
Comment (justinboswell) [inline include/aws/io/channel.h]: So near as I can tell, this mutex protects only this cross thread task queue, is that correct? Probably should note that here. Maybe even shove them in a struct together to make it super clear.

File: source/windows/iocp/iocp_event_loop.c [hunk 3]
```diff
@@ -429,6 +431,12 @@ static void s_schedule_task_now(struct aws_event_loop *event_loop, struct aws_ta
 static void s_schedule_task_future(struct aws_event_loop *event_loop, struct aws_task *task, uint64_t run_at_nanos) {
     s_schedule_task_common(event_loop, task, run_at_nanos);
 }
+
+static void s_cancel_task(struct aws_event_loop *event_loop, struct aws_task *task) {
+    struct iocp_loop *iocp_loop = event_loop->impl_data;
+    aws_task_scheduler_cancel_task(&iocp_loop->thread_data.scheduler, task);
+}
```
Comment (graebm) [inline source/windows/iocp/iocp_event_loop.c]: the task might also be in `iocp_loop->synced_data->tasks_to_schedule`

File: include/aws/io/channel.h [hunk 4]
```diff
@@ -59,6 +60,13 @@ struct aws_channel {
     struct aws_shutdown_notification_task shutdown_notify_task;
     aws_channel_on_shutdown_completed_fn *on_shutdown_completed;
     void *shutdown_user_data;
+    struct {
+        struct aws_linked_list pending_tasks;
+    } channel_thread_task_queue;
+    struct {
+        struct aws_mutex x_thread_task_lock;
+        struct aws_linked_list x_thread_pending_tasks;
+    } x_thread_task_queue;
```
Comment (graebm) [inline include/aws/io/channel.h]: debatable, trivial: naming suggestions
```
    struct {
        struct aws_linked_list list
    } channel_thread_tasks;
    struct {
        struct aws_mutex lock;
        struct aws_linked_list list;
    } cross_thread_tasks;
```

File: include/aws/io/channel.h [hunk 5]
```diff
@@ -79,6 +87,17 @@ struct aws_channel_slot {
...
+struct aws_channel_task;
+typedef void(aws_channel_task_fn)(struct aws_channel_task *channel_task, void *arg, enum aws_task_status status);
+
+struct aws_channel_task {
+    struct aws_task channel_task;
+    aws_channel_task_fn *user_task_fn;
+    void *task_data;
+    struct aws_linked_list_node node;
+    bool x_thread;
+};
```
Comment (graebm) [inline include/aws/io/channel.h]: debatable, trivial: naming suggestions
```
struct aws_channel_task {
    aws_channel_task_fn *task_fn;
    void *arg;
    struct aws_task wrapper_task;
    struct aws_linked_list_node node;
    bool is_cross_thread;
};
```

File: source/linux/epoll_event_loop.c [hunk 6]
```diff
@@ -288,6 +290,11 @@ static void s_schedule_task_future(struct aws_event_loop *event_loop, struct aws
     s_schedule_task_common(event_loop, task, run_at_nanos);
 }
 
+static void s_cancel_task(struct aws_event_loop *event_loop, struct aws_task *task) {
+    struct epoll_loop *epoll_loop = event_loop->impl_data;
+    aws_task_scheduler_cancel_task(&epoll_loop->scheduler, task);
```
Comment (graebm) [inline source/linux/epoll_event_loop.c]: the task might be in `epoll_loop->task_pre_queue` 😢

File: include/aws/io/channel.h [hunk 7]
```diff
@@ -147,6 +166,12 @@ struct aws_channel_handler {
 extern "C" {
 #endif
 
+/**
+ * Initializes channel_task for use.
+ */
+AWS_IO_API
+void aws_channel_task_init(struct aws_channel_task *channel_task, aws_channel_task_fn *task_fn, void *user_data);
```
Comment (graebm) [inline include/aws/io/channel.h]: trivial: it's called `arg` in task-town, not `user_data`

File: source/linux/epoll_event_loop.c (hunk 6 above)
Comment (JonathanHenson) [reply source/linux/epoll_event_loop.c]: separate task

**Pull request merged**